### ISSUE TITLE
feedback request for the client usage

### ISSUE THREAD

**Issue by EdJoPaTo** [bytebeamio/rumqtt#177]
Hi!
I've found this as a pure-rust implementation I can use for my needs of working with MQTT. But I am not sure if my approach is the intended one. I'm fairly new to rust and somehow found ways to work around my problems but maybe there are ways to do that way simpler. Also I want to provide an "outside view" how someone tried to use this lib without knowing much about it.

Currently I am building a Terminal UI in Rust similar to MQTT Explorer. See the repo [EdJoPaTo/mqttui](https://github.com/EdJoPaTo/mqttui).
The basic idea is to have three usecases:
- Simply publish something with two arguments: topic and payload
- Simply subscribe to something with 0-1 arguments topic (defaults to `#`)
- View a more detailed Terminal UI (`--interactive`) with an optionally different topic than `#`

I'm generally open to all kinds of feedback (as I said, I'm also fairly new to rust) but for this only three files are relevant:
- [mqtt_history.rs](https://github.com/EdJoPaTo/mqttui/blob/70040ffd051067011d9bb8b1b71c977ef3d49d28/src/mqtt_history.rs) for handling the event loop with the tui open and saving everything incoming to a HashMap
- [simple.rs](https://github.com/EdJoPaTo/mqttui/blob/70040ffd051067011d9bb8b1b71c977ef3d49d28/src/simple.rs) for handling the event loop on a simple publish or subscribe
- [main.rs](https://github.com/EdJoPaTo/mqttui/blob/70040ffd051067011d9bb8b1b71c977ef3d49d28/src/main.rs) for getting the specific stuff started and usage of the client (publish/subscribe/disconnect)

As I haven't used async rust yet I went for the synchronous client/example. Also as I know now the [tui crate](https://crates.io/crates/tui) is also synchronous and would need to create another thread in order to use anything async. So I can just use what the Sync wrapper of this lib already does.

Using this library sparked a few questions:
- As far as I understand MQTT you do not always need a client name. For example the "simply publish something once" task wouldn't really need a client name? Currently I just randomly create something in order to have multiple clients running at the same time.
- As the synchronous event loop is blocking I have to get the loop to a background thread in order to do something independently. For example when the hostname is wrong and can not be connected to the thread panics but the main thread is continuing without any problems. When I try to join the thread handle I would get that "something failed" but as long as the TUI is open the user just wouldn't notice. I kinda solved that with [doing one iteration of the loop before spawning a thread](https://github.com/EdJoPaTo/mqttui/blob/70040ffd051067011d9bb8b1b71c977ef3d49d28/src/mqtt_history.rs#L18-L20). Then stuff like wrong hostnames will fail immediately. For all other things I always just [get the mutex lock](https://github.com/EdJoPaTo/mqttui/blob/70040ffd051067011d9bb8b1b71c977ef3d49d28/src/mqtt_history.rs#L34-L36). If something fails then the mutex gets poisoned and the main thread will notice the poisoned mutex. That is somehow working but these workarounds feel wrong. Whats the idea of using it in a better way?
- If the synchronous event loop would not be blocking I could just try to advance the event loop with every tick I use to update the tui. But I guess its not as easy to do with the synchronous client just being a wrapper around the AsyncClient?
- In case of the simple publish [I wait in the event loop for getting a PubAck and then call client.disconnect()](https://github.com/EdJoPaTo/mqttui/blob/70040ffd051067011d9bb8b1b71c977ef3d49d28/src/simple.rs#L18-L21). This somehow mixes the event loop and the client. I havnt read much into the AsyncClient but as far as I understand the `client.publish(…).await` call would be waiting until the Ack? So with an AsyncClient this code would be nicer? In case I wouldn't have the synchronous tui (I could also work around): A user of the lib should just prefer the AsyncClient? Or am I missing something there?
  If that's the case I would like to see a hint somewhere that the synchronous way is possible but just a wrapper around the async one. At the start of getting into the lib I assumed it more like a "either or" and not as a "direct or indirect but still the same".
- [When the event loop notices an outgoing disconnect I end the loop](https://github.com/EdJoPaTo/mqttui/blob/70040ffd051067011d9bb8b1b71c977ef3d49d28/src/simple.rs#L13-L15). Not doing that would result in a connection error as the broker closes the connection. Is that just the way to go?

I hope I can learn more about how to use this library in a better way. Thanks in advance!

**Comment by tekjar** [comment]
> * As far as I understand MQTT you do not always need a client name. For example the "simply publish something once" task wouldn't really need a client name? Currently I just randomly create something in order to have multiple clients running at the same time.

Yeah. You need name for persistent sessions though. Most of the uses cases I see uses names. It's just that I didn't bother implementing it as I don't think there is any drawback.
 

> * For example when the hostname is wrong and can not be connected to the thread panics but the main thread is continuing without any problems. When I try to join the thread handle I would get that "something failed" but as long as the TUI is open the user just wouldn't notice. I kinda solved that with [doing one iteration of the loop before spawning a thread](https://github.com/EdJoPaTo/mqttui/blob/70040ffd051067011d9bb8b1b71c977ef3d49d28/src/mqtt_history.rs#L18-L20).

This question boils down to "How do we distinguish configuration related failures like bad hostname from robustness related failures like broker going down or bad network". Answer is we can't. The way I would solve this is

- `iter` loop on current thread till `ConnAck` notification is received. This implies that configuration is fine. 
- send `Connection` to background thread

There used to be `connection.connect()` which does `wait till connack success` but I removed it as it might give an intuition that you have to call `connect` on every failure.

But there are other configuration related errors that can happen much later after the initial connection is successful. Like what do you do when client certificate expires or dns/ip has changed? How you deal with these is very use-case specific :)

> * For all other things I always just [get the mutex lock](https://github.com/EdJoPaTo/mqttui/blob/70040ffd051067011d9bb8b1b71c977ef3d49d28/src/mqtt_history.rs#L34-L36). If something fails then the mutex gets poisoned and the main thread will notice the poisoned mutex. That is somehow working but these workarounds feel wrong. Whats the idea of using it in a better way?

I usually use channels. 

> * In case of the simple publish [I wait in the event loop for getting a PubAck and then call client.disconnect()](https://github.com/EdJoPaTo/mqttui/blob/70040ffd051067011d9bb8b1b71c977ef3d49d28/src/simple.rs#L18-L21). This somehow mixes the event loop and the client.

Why do you `disconnect` after every publish? There is no functional difference between async and sync. Async lets you easily multiplex with other async code for concurrent execution. If you can afford a separate thread, sync client is easier

> * I havnt read much into the AsyncClient but as far as I understand the `client.publish(…).await` call would be waiting until the Ack?

No it doesn't. This is an asynchronous client written for performance. 'Asynchronous' as in 'next publish doesn't wait for previous ack' and not to be confused with asynchronous rust.

> * [When the event loop notices an outgoing disconnect I end the loop](https://github.com/EdJoPaTo/mqttui/blob/70040ffd051067011d9bb8b1b71c977ef3d49d28/src/simple.rs#L13-L15). Not doing that would result in a connection error as the broker closes the connection. Is that just the way to go?

I need to fix this. If there is an error after `disconnect`, connection should return `Error::Shutdown` for users to break the loop. It'll help if you can create a separate issue for this

**Comment by tekjar** [comment]
@EdJoPaTo Also I have similar terminal UI plans [here](https://github.com/bytebeamio/mqttwrk). We can join efforts if you are interested

**Comment by EdJoPaTo** [comment]
> Why do you disconnect after every publish?

The idea is to send a single publish via command line. Publish the payload on a topic, wait for ack, disconnect and the command is finished.
Mainly just publish it and stay open long enough to be sure its done.

That is also a case where I don't think a client name for persistence has any use as its something happening once.

> Also I have similar terminal UI plans here. We can join efforts if you are interested

Currently my client is mixing a command line interface working with stdout and the rich Terminal UI.
mqttwrk intention seems to be a (benchmarking) cli tool. Splitting this into a command line interface from "just subscribe this and write to stdout" or "just publish this once" to "benchmark how many you can publish" and a terminal ui seems like a good idea.
Not sure how well I might be able to help there.

---

Another thing: Having to handle the event loop and do things like "break the loop when Disconnect was sent" myself is great in order to have control but doesn't really feel like a "high level" client to me yet. For something high level I would assume 2 things:
- Some way where subscribed packages arrive. Maybe a Channel rx, maybe an iterator I can try_next() or something like that?
- Some methods I can use to "control" something like the existing client already does. publish something, subscribe something, shutdown and everything ends, …. Which also have some kind of Result<Success, Error> which tells me that something failed (on start up / bad network / …). Like when I want to publish something I would like an Result for that. Did that work? Currently that comes up in a different place, the event loop.

Don't get me wrong, sometimes its awesome to have control over all the things happening but for "just use it" I am happily accepting some defaults and wont care how exactly some event loop is doing what I want in the background.

For example for the simple "just publish something once and than end" I want to have. Currently I have to ensure to work correctly with the event loop to ensure my "client.publish()" is actually sent out. I have to understand it which is something I wouldn't expect from a high level client.

**Comment by tekjar** [comment]
> Having to handle the event loop and do things like "break the loop when Disconnect was sent" myself is great in order to have control but doesn't really feel like a "high level" client to me yet.

Maybe. But that's a trade off because of 'automatic reconnections' feature. I automatic reconnections a necessary feature for a highlevel client? The answer depends on who you are asking.

> * Some way where subscribed packages arrive. Maybe a Channel rx, maybe an iterator I can try_next() or something like that?

I think the current behaviour is not very different from this? `connection.iter().next()`?  Instead of just incoming packets it yields both incoming and outgoing. Match on just incoming? 

But I think I understand all your concerns. You don't want to deal with anything other than 
- Is my publish or subscribe successful
- For incoming messages, my iterator/channel shouldn't yield anything other than incoming publish.

I think these points are suggesting towards a synchronous client. I.e 
- `client.publish()` should block until an ack is received. This is a performance killer. This library is mainly written keeping vehicles (which transmit and receive data in realtime) as primary usecase (high message rate and dynamic network conditions). 

I personally don't have time to implement synchronous client. But I'm happy to support if someone is ready to take that task.

**Comment by EdJoPaTo** [comment]
> I think these points are suggesting towards a synchronous client. I.e
>
> - client.publish() should block until an ack is received. This is a performance killer.

From what I think so far about asynchronous Rust I assume this synchronous client will work with async Rust. Calling Publish would publish and return a future. Then let something else do its work until its `wake()` is called when the Ack was received.
As all these tasks then are able to run asynchronous it shouldnt be a performance killer. (Coming from NodeJS where its basically a single core async engine: It can do a lot while still being JavaScript. There are differences in the away how async await works but the basic idea is the same)

I don't think its that useful for the current TUI projects but for another project I have in mind I will take a look on how to achieve something like that. I have some different ideas what might work and how much of a synchronous client would actually be helpful (to me). (Whenever I have time for that…)

Thanks for your feedback! I was able to improve some things and get a better understanding of the ideas why it was build like that. :)

**Comment by tekjar** [comment]
> From what I think so far about asynchronous Rust I assume this synchronous client will work with async Rust.

Ofcourse. It's also easy enough. One possible implementation is
- Save oneshot channel along with publish. 
- `client.publish()` will be blocked on oneshot channel
- Fire or drop the channel after the ack

> Calling Publish would publish and return a future. Then let something else do its work until its `wake()` is called when the Ack was received.

If you have some other IO to schedule while waiting for ack, probably yes. But vehicle's or drone's `data pipeline` is still going to be slow.

**Comment by EdJoPaTo** [comment]
After I worked a bit more with the library its helpful to have it this way with the event loop. Its relatively easy to do specifically whats needed for the project. Which differs more often than I first expected. Sometimes a dedicated thread with channels is a good way, sometimes it can be just done as a simple method call from the event loop.

Not sure if I would still really need a more synchronous client (publish something and be sure it worked without errors). It was more what I was used to before.

Thanks for the library again, after getting used to it it works well for me in multiple cases :)
